### PR TITLE
Issue 2732: Fix serial number and expiration

### PR DESCRIPTION
**Pull request by wiktor-k**
This is a mini PR to fix issue #2732.

Fortunately serial numbers have the same exact length as OpenPGP fingerprints v4 so it looks nice.

------------

// todo:
 - [ ] test for fingerprint / serial number
 - [ ] test expiration

### CODE REVIEW

**Comment by wiktor-k** [comment]
> send with broken S/MIME cert - err
> --
> ℹ (attempt 1 of 3) Retrying: Error: this.wait_till_gone: some of "@dialog-add-pubkey" still present after timeout:20

:thinking: hmm I guess it fails because the add pubkey dialog is not dismissed and that's because there's some error triggered when adding the key (maybe related to expiration?). Is there a way to get more details out of it? (running locally fails with missing `test-secrets.json` -> of course :man_facepalming: ).

**Comment by tomholub** [comment]
> > ## send with broken S/MIME cert - err
> > information_source (attempt 1 of 3) Retrying: Error: this.wait_till_gone: some of "@dialog-add-pubkey" still present after timeout:20
> 
> thinking hmm I guess it fails because the add pubkey dialog is not dismissed and that's because there's some error triggered when adding the key (maybe related to expiration?). Is there a way to get more details out of it? (running locally fails with missing `test-secrets.json` -> of course man_facepalming ).

Unsure about why this happened - what you said sounds plausible (I can look into it tomorrow).

But yes - you can get more details. It will spam your email with details for every failure regardless of who pushes the commit. On my email I filter them and route them all straight to trash, and if I need a particular result, I'll search for it (which searches trash too for up to 30 days on Gmail).

I'll add your email there and re-run, so you should receive it.

**Comment by tomholub** [comment]
I've re-run it: https://flowcrypt.semaphoreci.com/jobs/e7795d41-0d9b-429b-be4f-bf68fb9013ff

It failed again, and you should receive a debug email. Over time, you'll be getting loads of these emails (sometime with a bit of a delay), and so you'll need a way to search them. For this, take a look at the FAIL ID:

image


Here it was `r2r4i1i0gjr`. Search in Gmail:

image

The attachment is a HTML file. Download and open it:

Click on the collapsed red part:

image

image

**Comment by tomholub** [comment]
As for the failure, maybe we click it too fast? Let me try to add a delay somewhere.

**Comment by tomholub** [comment]
Ah! The problem is that now the faulty cert gets discovered sooner than before.

Which is a good thing.

That means we have to start looking for the error sooner.

These kinds of issues would be easier for you to debug locally. Anyway, I'll go ahead and fix it.

File: extension/js/common/core/smime.ts [hunk 1]
```diff
@@ -16,3 +16,26 @@ export const encrypt = (pubkeys: string[], data: Uint8Array): { data: Uint8Array
...
 };
+
+export const getSerialNumber = (pubkey: string) => {
+  const certificate = forge.pki.certificateFromPem(pubkey);
+  return certificate.serialNumber;
+};
+
+export const getExpirationTime = (pubkey: string) => {
+  const certificate = forge.pki.certificateFromPem(pubkey);
+  const expiration = certificate.validity.notAfter;
```
Comment (tomholub) [inline extension/js/common/core/smime.ts:27]: It's preferred to return a `Date` for dates, so in this case `Date | undefined` would be ideal

File: extension/js/common/platform/store/contact-store.ts [hunk 2]
```diff
@@ -105,24 +106,23 @@ export class ContactStore extends AbstractStore {
       }
       // X.509 certificate
       if (PgpKey.getKeyType(pubkey) === 'x509') {
-        // FIXME: For now we return random data.
-        // Later we'll return serial ID from the certificate.
-        const longid = Math.random() + '';
+        const fingerprint = smime.getSerialNumber(pubkey);
+        const longid = fingerprint.substring(fingerprint.length / 2);
```
Comment (tomholub) [inline extension/js/common/platform/store/contact-store.ts:110]: Is this following any known S/MIME spec or just trying to shoehorn it into our OpenPGP code?

File: extension/js/common/platform/store/contact-store.ts [hunk 3]
```diff
@@ -105,24 +106,23 @@ export class ContactStore extends AbstractStore {
...
-          fingerprint: Math.random() + '',
+          fingerprint,
           longid,
           longids: [longid],
           pending_lookup: 0,
           last_use: lastUse || null,
           pubkey_last_sig: lastSig || null,
           pubkey_last_check: lastCheck || null,
-          expiresOn: null
+          expiresOn: smime.getExpirationTime(pubkey) || null
```
Comment (tomholub) [inline extension/js/common/platform/store/contact-store.ts:125]: While the Date will be a bit less convenient here (need to convert to number), I still think it's generally better to work with a Date.

File: extension/js/common/ui/key-import-ui.ts [hunk 4]
```diff
@@ -146,7 +147,7 @@ export class KeyImportUi {
   }
 
   public checkPub = async (armored: string): Promise<string> => {
-    if (PgpKey.getKeyType(armored) === 'x509') {
+    if (PgpKey.getKeyType(armored) === 'x509' && smime.isEmailCertificate(armored)) {
```
Comment (tomholub) [inline extension/js/common/ui/key-import-ui.ts:150]: I would make the error a bit more obvious, pseudo code
```
if(isx509)
  if(!isEmailCert) {
    throw UserAlert('The type of this x509 certificate is unknown.\nEither provide a valid S/MIME certificate, or an OpenPGP public key.');
  }
  return armored;
}
```

**Comment by tomholub** [comment]
You could then also add unit tests, to test that it produces expected serial id (compared to some 3rd party s/mime software), "longid" and also test one key that expires, that it yields the right expiration and one that doesn't expire.

You'd add them here: https://github.com/FlowCrypt/flowcrypt-browser/blob/master/test/source/tests/tests/unit.ts

Unit tests shouldn't really need test secrets, I think. But I'm not sure if they run without them.

**Comment by wiktor-k** [comment]
Got it! I think I'll roll this expiration and serial ID thing into one class so that cert parsing will happen only once. I'm planning that after I finish the introduction of proper key type (with openpgp / x.509 variants) but changing that has ripple effects over the entire codebase so I'm slowly progressing on that front :)

**Comment by tomholub** [comment]
@wiktor-k what is the status here - is it waiting for my action?

**Comment by wiktor-k** [comment]
No, I'm planning to include this as soon as the refactoring thing is done (see there is a `TODO` in the refactoring to include this change too). No worries if I'll need your confirmation I'll explicitly ask.

**Comment by wiktor-k** [comment]
Changes from this PR are already included in #2739 so closing this one.

**Pull request closed without merging**